From patchwork Thu Jan 17 19:55:58 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Jim Mattson <jmattson@google.com>
X-Patchwork-Id: 10768909
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 93D9313BF
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 17 Jan 2019 19:56:10 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 84DB82AA59
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 17 Jan 2019 19:56:10 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 78D202AE00; Thu, 17 Jan 2019 19:56:10 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 2253C2AA59
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 17 Jan 2019 19:56:10 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1729066AbfAQT4I (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Thu, 17 Jan 2019 14:56:08 -0500
Received: from mail-yb1-f202.google.com ([209.85.219.202]:56066 "EHLO
        mail-yb1-f202.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1728797AbfAQT4I (ORCPT <rfc822;kvm@vger.kernel.org>);
        Thu, 17 Jan 2019 14:56:08 -0500
Received: by mail-yb1-f202.google.com with SMTP id b13so4566348ybr.22
        for <kvm@vger.kernel.org>; Thu, 17 Jan 2019 11:56:08 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:message-id:mime-version:subject:from:to:cc;
        bh=xwot+Gh8ndi4kmsgDNI9FP8kjiLu0s25KsF8UNnCaRA=;
        b=aTYFo76xzX/6itXx+z/Ba+lvgRkHfiXFGFnOlSSkn/ZJ7a1IpcC2Tswoi8s1IuzOJt
         +1mFxi/jkHLdMX05gZXOTg5S27E/jmjKPjBm4YHK/G70boEC/rPMofvtdk49CHrAWzhm
         8ZtS51x74TzxbNdU4K+t6YUSCvlpRHGSH0ggiI7XMLf4lS/kd0HwXYTsRqjno8v3GJgQ
         Ton+Iv0ftOBbScLQ95277cMXSVO1AhAphRl7fSCu8ig4vkNmgokh1Ft8YbQwDl/SgdPm
         oEAHAPNMOrRshxdR8CvE6pjEfUq7n2k6QNjprKW9sjywFZqhkFoVgzJ9kLTi/6nsb1Jl
         ep6w==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:message-id:mime-version:subject:from:to:cc;
        bh=xwot+Gh8ndi4kmsgDNI9FP8kjiLu0s25KsF8UNnCaRA=;
        b=N/SQ3Den0OScBxHMvE6srcJlH4zGMabu3oxDKx/Yj/osIj9SdIx8K7uXvfySfIJzdT
         WqRCSgLj/dqql2YuDSmhbMtUaXGB+zaDwNIAKsnkwrcSDRfwy93M+yRp+r+lWFOsrfxF
         UpJi3vQHm7erh8qp6FcI2zm8m2XCje4K+SMhK0fED6rnPCaGJDbkz4bu94gfL+hA8ZBa
         nTUWD/LXRH8Gbc3jUaWhsvfUROWwDwJrMAcCW1Ac9zXDjGgHeCYBPlf9fXo0c7kYA/Ou
         fSdY8ONSYp63HZmZTMF+EI3lwaiD0EjY4vBbKd6ppEdcuZbK7xEdE6/TjoKOzBa2fchE
         SB3w==
X-Gm-Message-State: AJcUukdZ964ZLMzarDduY8mAjXwKYJug2YTK/RYKgE/MFSdqHh31q+jr
        49d89TMpcI0howwp6gu5gQtB9FTM235vM5lRYCLD5hse7RqrYtclaEfyYQ9y+sfJxTG1otpsazt
        9mwPS/kth0yvrQsVg1k+bKHfldAmUNZMytz5ALMKLPMnpsnYM4xBQX3+QOHpTKSc=
X-Google-Smtp-Source: 
 ALg8bN4nIxzVfSMP1V7zTzTIPOVyMJI+B5dlVwM1Lo+YVe99B3bvIBhRR0RtwB6T+k/vlxUrtS1cBhUReS5OVA==
X-Received: by 2002:a25:bb01:: with SMTP id z1mr1880311ybg.83.1547754967560;
 Thu, 17 Jan 2019 11:56:07 -0800 (PST)
Date: Thu, 17 Jan 2019 11:55:58 -0800
Message-Id: <20190117195558.110516-1-jmattson@google.com>
Mime-Version: 1.0
X-Mailer: git-send-email 2.20.1.97.g81188d93c3-goog
Subject: [PATCH] KVM: nVMX: Fix size checks in vmx_set_nested_state
From: Jim Mattson <jmattson@google.com>
To: kvm@vger.kernel.org
Cc: Jim Mattson <jmattson@google.com>,
        Felix Wilhelm <fwilhelm@google.com>,
        Drew Schmitt <dasch@google.com>, Marc Orr <marcorr@google.com>,
        Peter Shier <pshier@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

The size checks in vmx_nested_state are wrong because the calculations
are made based on the size of a pointer to a struct kvm_nested_state
rather than the size of a struct kvm_nested_state.

Reported-by: Felix Wilhelm  <fwilhelm@google.com>
Signed-off-by: Jim Mattson <jmattson@google.com>
Reviewed-by: Drew Schmitt <dasch@google.com>
Reviewed-by: Marc Orr <marcorr@google.com>
Reviewed-by: Peter Shier <pshier@google.com>
---
 arch/x86/kvm/vmx/nested.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/arch/x86/kvm/vmx/nested.c b/arch/x86/kvm/vmx/nested.c
index 2616bd2c7f2c..3bb49ad91d0c 100644
--- a/arch/x86/kvm/vmx/nested.c
+++ b/arch/x86/kvm/vmx/nested.c
@@ -5351,7 +5351,7 @@ static int vmx_set_nested_state(struct kvm_vcpu *vcpu,
 		return ret;
 
 	/* Empty 'VMXON' state is permitted */
-	if (kvm_state->size < sizeof(kvm_state) + sizeof(*vmcs12))
+	if (kvm_state->size < sizeof(*kvm_state) + sizeof(*vmcs12))
 		return 0;
 
 	if (kvm_state->vmx.vmcs_pa != -1ull) {
@@ -5395,7 +5395,7 @@ static int vmx_set_nested_state(struct kvm_vcpu *vcpu,
 	    vmcs12->vmcs_link_pointer != -1ull) {
 		struct vmcs12 *shadow_vmcs12 = get_shadow_vmcs12(vcpu);
 
-		if (kvm_state->size < sizeof(kvm_state) + 2 * sizeof(*vmcs12))
+		if (kvm_state->size < sizeof(*kvm_state) + 2 * sizeof(*vmcs12))
 			return -EINVAL;
 
 		if (copy_from_user(shadow_vmcs12,
